internal/http/client.go

```diff
@@ -40,31 +52,93 @@ func (client Client) AppendDefaultHeaders(headers RequestHeaders) RequestHeaders
 	return headers
 }
 
-// Do does a request
-func (client Client) Do(method string, url string, headers RequestHeaders, body io.Reader) ([]byte, ResponseHeaders, error) {
-	if client.limiter != nil {
-		ctx := context.Background()
-		client.limiter.Wait(ctx)
+func (c Client) retryDuration(retry int) time.Duration {
+	duration := c.retryBaseDuration * (1 << retry)
+	if duration > c.retryMaxDuration {
+		return c.retryMaxDuration
 	}
 
-	headers = client.AppendDefaultHeaders(headers)
+	duration += time.Duration(rand.Float64() * float64(time.Second))
 
-	req, err := http.NewRequest(method, url, body)
-	if err != nil {
-		return nil, ResponseHeaders{}, err
+	return duration
+}
+
+// Do does a request
+func (c Client) Do(method string, url string, headers RequestHeaders, body io.Reader) ([]byte, ResponseHeaders, error) {
+	var requestBody []byte
+	var err error
+
+	if body != nil {
+		requestBody, err = io.ReadAll(body)
+		if err != nil {
+			return nil, ResponseHeaders{}, err
+		}
 	}
 
-	for _, header := range headers {
-		req.Header.Set(header.Name, header.Value)
+	if c.limiter != nil {
+		ctx := context.Background()
+		c.limiter.Wait(ctx)
 	}
 
-	resp, err := client.Client.Do(req)
-	if err != nil {
-		return nil, ResponseHeaders{}, err
+	headers = c.AppendDefaultHeaders(headers)
+
+	var resp *http.Response
+
+	for i := 0; i < c.retryCount+1; i++ {
+		req, err := http.NewRequest(method, url, bytes.NewReader(requestBody))
+		if err != nil {
+			return nil, ResponseHeaders{}, err
+		}
+
+		for _, header := range headers {
+			req.Header.Set(header.Name, header.Value)
+		}
+
+		resp, err = c.client.Do(req)
+		if err != nil {
+			// If no need to retry, return early
+			if i == c.retryCount {
+				return nil, ResponseHeaders{}, err
+			}
+
+			// Retry using exp backoff
+			time.Sleep(c.retryDuration(i))
+			req.Body = io.NopCloser(bytes.NewReader(requestBody))
+
+			continue
+		}
+
+		defer resp.Body.Close()
+
+		// Break if client config not set for retries or retries maxed out
+		if c.retryCount == i {
+			break
+		}
+
+		isRetryResponse := resp.StatusCode == 429 || resp.StatusCode > 500
+		if !isRetryResponse {
+			break
+		}
+
+		var waitTime time.Duration
+
+		// Use Shopify's retry after duration if any
+		if retryAfterHeader := resp.Header.Get("Retry-After"); retryAfterHeader != "" {
+			retryAfter, _ := strconv.ParseFloat(retryAfterHeader, 64)
+			waitTime = time.Duration(retryAfter) * time.Second
```

Comment:
```suggestion
			waitTime = time.Duration(retryAfter * float64(time.Second))
```
A `Retry-After` header of `2.9` seconds is currently becoming `2.0` seconds since it's being converted to an int before multiplying the `time.Second` constant

Reply:
oh yeah is an int64, nice catch